cmake/ProhibitInSourceBuild.cmake

```diff
@@ -0,0 +1,23 @@
+
+# Distributed under the University of Illinois/NCSA Open Source License.
+# See LICENSE.txt for details.
+
+function(prohibit_in_source_build)
+  if ("${CMAKE_BINARY_DIR}" STREQUAL "${CMAKE_SOURCE_DIR}")
```

Comment:
I think we should also prohibit building in `src`, `cmake`, `tests`. I don't think we need to explicitly block every directory in our source dir, but I could see someone accidentally building in `src` for example. Thoughts?